### ISSUE TITLE
Add a method to search across multiple columns

### ISSUE THREAD

**Issue by dshukertjr** [supabase/postgrest-js#289]
# Feature request

## Is your feature request related to a problem? Please describe.

Currently, `textSearch` allows to search for a search query on only one column, but in a typical application, we would want to search across multiple columns. 

Currently, searching through multiple columns is possible using the rest api like this https://github.com/PostgREST/postgrest/issues/2374#issuecomment-1186255788

```http
GET /blog_posts?or=(title.fts.searchTerm,content.fts.searchTerm)
```
That means we can perform full text search across multiple columns via js client can be done like this:
```typescript
const { data, error } = await supabase
  .from('blog_posts')
  .select('*')
  .or('title.fts.searchTerm,content.fts.searchTerm')
```

However, developers would need to use the rest api syntax to perform text search, so it might not be a nice developer experience. 

## Describe the solution you'd like

This is a suggestion from @steve-chavez, so I don't want to take any credits, but we could introduce a new method where we combine multiple text search just like `match` like this:

```typescript
const { data, error } = await supabase
  .from('blog_posts')
  .select('*')
  .textMatch({title: 'sometitle', contents: 'somecontent'})
```

## Describe alternatives you've considered

We can use `or` filter to combine multiple text search query.
```typescript
const { data, error } = await supabase
  .from('blog_posts')
  .select('*')
  .or('title.fts.searchTerm,content.fts.searchTerm')
```

**Comment by soedirgo** [comment]
Sounds reasonable to me. Simplest solution would be to extend the current `textSearch()` to accept an array of columns for the `column` argument:
```ts
const { data, error } = await supabase
  .from('blog_posts')
  .select()
  .textSearch(['title', 'contents'], 'searchTerm', { config: 'english' })
```

**Comment by soedirgo** [comment]
Hmm, it departs too much from how the current filters look like imo (personal taste of course). Though I can see the confusion of `or`/`and` with the array approach.

**Comment by dshukertjr** [comment]
I actually like @soedirgo's suggestion here, because I think in most cases, we need to search for only one search query across different columns. About the confusion of `or` and `and`, would in code documentation be not enough?

```typescript
const { data, error } = await supabase
  .from('blog_posts')
  .select()
  .textSearch(['title', 'contents'], 'searchTerm')
```

**Comment by steve-chavez** [comment]
> I think in most cases, we need to search for only one search query across different columns

Yeah, since it's a common use case for `textSearch` then I think it should be fine adding a shortcut for it.

**Comment by steve-chavez** [comment]
This [tweet](https://twitter.com/cawthornejosh/status/1549362423689355265?s=21&t=kj_Nsq_nw4kCwsob-lUr3g) asks for something similar for the `like` operator. I guess like:

```js
const { data, error } = await supabase
  .from('blog_posts')
  .select()
  .ilike(['title', 'contents'], 'Jo%')
```

**Comment by riodw** [comment]
Any updates on this? I need it Lol

**Comment by riodw** [comment]
Might I suggest:
```
const { data, error } = await supabase
  .from('books')
  .select()
  .textSearch(`'title' | 'description'`, `'little' | 'big'`);
```
This is a simple modification of https://supabase.com/docs/guides/database/full-text-search#match-any-search-words

**Comment by kvetoslavnovak** [comment]
I agree that the lack of this feature is really really annoying. To use an SQL function for this is rather overkill.
There are already  SackOverflow  [questions](https://stackoverflow.com/questions/69225396/is-there-a-way-to-perform-full-text-search-on-multiple-columns-on-supabase-with)  and even otherwise nice examples in official docs [do not explicitly mention and warn](https://supabase.com/docs/guides/database/full-text- search#searchable-columns) that there is no JavaScript and Dart front-end api for more columns search.

**Comment by shrey150** [comment]
Building a project where this feature would be greatly appreciated. Might use the workaround approach with a custom Postgres function, but this really should be a first-class function in the API

**Comment by steve-chavez** [comment]
Just to document how you can do this:

```sql
CREATE FUNCTION title_description(books) RETURNS text AS $$
  SELECT $1.title || ' ' || $1.description;
$$ LANGUAGE SQL;
```

```js
const { data, error } = await supabase
  .from('books')
  .select()
  .textSearch('title_description', `little`)
```

https://supabase.com/docs/guides/database/full-text-search#search-multiple-columns

**Comment by lucksp** [comment]
how do you all accomplish this now, without the new feature?  Use multiple queries and then add logic to filter/combine?

**Comment by Strijdhagen** [comment]
> Just to document how you can do this now:
> 
> ```sql
> CREATE FUNCTION title_description(books) RETURNS text AS $$
>   SELECT $1.title || ' ' || $1.description;
> $$ LANGUAGE SQL;
> ```
> 
> ```js
> const { data, error } = await supabase
>   .from('books')
>   .select()
>   .textSearch('title_description', `little`)
> ```
> 
> https://supabase.com/docs/guides/database/full-text-search#search-multiple-columns

@steve-chavez is it possible that this doesn't work for a materialized view?
I'm getting the following error:

code
: 
"42703"
details
: 
null
hint
: 
null
message
: 
"column search_1.name_slug does not exist"

**Comment by lucksp** [comment]
any update on this feature being added to roadmap &/or implemented?

**Comment by bensontrent** [comment]
@lucksp Thanks a lot for that tip. Your solution works beautifully for returning partial text and multiple word searches across multiple columns.  I modified your function to allow for selecting by account and test mode and pagination. Here's what I came up with:

My SQL Function

``` sql
CREATE OR REPLACE FUNCTION fuzzy_search_shipments(
    term TEXT,
    account_id TEXT,
    is_test_mode BOOLEAN,
    start_index INTEGER,
    end_index INTEGER
)
RETURNS SETOF shipments AS $$
BEGIN
    RETURN QUERY
    SELECT
        s.*
    FROM
        shipments s
    WHERE
        s.account_id = fuzzy_search_shipments.account_id
    AND
        s.is_test_mode = fuzzy_search_shipments.is_test_mode
    AND
        (
            similarity(s.to_email, term) > 0.2
        OR
            similarity(s.to_company, term) > 0.2
        OR
            similarity(s.to_name, term) > 0.2
        )
    ORDER BY
        s.created_at DESC
    LIMIT (end_index - start_index + 1)
    OFFSET start_index;
END;
$$ LANGUAGE plpgsql;
```

In my Typescript I request one more row than specified by the `pageSize` and `page` so I can provide a `has_more` `boolean` in my response.  If the returned results returns more than the length of rows specified, I remove the last row and return the response with the `has_more` Boolean set to `true`.  Is for an express.js app
``` js
  async fuzzySearch(req: Request, res: Response) {
    const page = req.query.page ? parseInt(req.query.page + '') : 1; // Get page number from query parameter
    const pageSize = req.query.pageSize ? parseInt(req.query.pageSize + '') : 10; // Set the number of items per page
    const searchTerm = req.query.searchTerm ? (req.query.searchTerm + '').toLowerCase() : ''; // Get search term from query parameter
    const mode = req.query.mode === 'test' ? true : false;

    try {
        if (supabase && accountId) {
            const { data: shipments, error } = await supabase
            .rpc('fuzzy_search_shipments', {
              term: searchTerm,
              account_id: accountId as string,
              is_test_mode: mode,
              start_index: (page - 1) * pageSize,
              end_index: page * pageSize,
            })

            if (error) {
                throw error.message;
            }

            let has_more = false;

            if (shipments && shipments.length > pageSize) {
                has_more = true;
                shipments.pop();

            }

            res.send({
                has_more: has_more,
                data: shipments,
                searchTerm: req.query.searchTerm
            });
        }

    } catch (error) {
        console.error(error);
        res.status(500).send({
            error: error,
            data: null
        });
    }
```
